File: MobileAcceleratorExample/WebVC.m
//
//  WebVC.m
//  MobileAcceleratorExample
//
//  Created by Brian Salomon on 9/16/15.
//  Copyright © 2015 Akamai. All rights reserved.
//

#import "WebVC.h"

@interface WebVC ()

@property (nonatomic, strong) IBOutlet UIWebView *webView;
@property (nonatomic) NSTimeInterval startTime;
@property (nonatomic) NSTimeInterval loadTimeElapsed;
@property (nonatomic, strong) UIAlertController *alertController;

@end

@implementation WebVC

- (void) viewDidLoad {
    [super viewDidLoad];

    self.automaticallyAdjustsScrollViewInsets = NO;

    self.startTime = [NSDate timeIntervalSinceReferenceDate];

    NSURLRequest *urlRequest = [NSURLRequest requestWithURL:self.url
                                                cachePolicy:NSURLRequestUseProtocolCachePolicy
                                            timeoutInterval:60.0];
    [self.webView loadRequest:urlRequest];
}

- (void) viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];

    if ([self.webView isLoading]) {
        [self.webView stopLoading];
    }
}

- (void) dealloc
{
    self.webView.delegate = nil;
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
}

- (void) flashMessage:(NSString *)message withTitle:(NSString*)title
{
    if (self.alertController) {
        // alert already showing.  try again soon.
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self flashMessage:message withTitle:title];
        });
        return;
    }

    if (! [self.view superview]) {
        NSLog(@"Left webview.  Skipping title: %@, message: %@", title, message);
        return;
    }

    self.alertController = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleActionSheet];

    if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
        CGRect bounds = self.view.bounds;
        CGFloat bottomY = bounds.origin.y + bounds.size.height;
        bounds = CGRectMake(bounds.origin.x, bottomY, bounds.size.width, 0.0);
        self.alertController.popoverPresentationController.sourceRect = bounds;
        self.alertController.popoverPresentationController.sourceView = self.view;
    }

    [self presentViewController:self.alertController animated:YES completion:^{
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            [self dismissViewControllerAnimated:YES completion:^{
                self.alertController = nil;
            }];
        });

    }];
}

#pragma mark UIWebViewDelegate

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    // Note: we would like to clear network stats before each new page,
    //       but this is called several times (once per frame).
    //       Going beyond this demo, one could accumulate stats across
    //       frames and store those stats for each URL loaded.
}

- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    NSTimeInterval timeNow = [NSDate timeIntervalSinceReferenceDate];
    if (! webView.isLoading) {
        self.loadTimeElapsed = timeNow - self.startTime;

        [self flashMessage:self.url.absoluteString withTitle:[NSString stringWithFormat:@"%.2fs", self.loadTimeElapsed]];

        [self.webDelegate loadedWebSiteName:[webView.request.mainDocumentURL absoluteString]];
    }
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
    [self flashMessage:@"Cannot load page" withTitle:@"Offline"];
}

@end

File: MobileAcceleratorExampleSwift/WebVC.swift
//
//  webView.swift
//  MobileAcceleratorExampleSwift
//
//  Created by Waazim Reza on 23/05/2016.
//  Copyright © 2016 Akamai. All rights reserved.
//

import Foundation


protocol WebVCDelegate: class {
	func loadedWebSiteName(siteName: String)
}

final class WebVC: UIViewController, UIWebViewDelegate {

	var siteName = ""
	var url: URL?
	weak var webDelegate: WebVCDelegate?

	@IBOutlet weak var webView: UIWebView!
	private var startTime: TimeInterval = 0.0
	private var loadTimeElapsed: TimeInterval = 0.0
	private var alertController: UIAlertController!

	override func viewDidLoad() {
		super.viewDidLoad()

		automaticallyAdjustsScrollViewInsets = false

		guard let url = url else {
			print("URL not set")
			return
		}

		startTime = Date.timeIntervalSinceReferenceDate

		let urlRequest = URLRequest.init(url: url, cachePolicy:.useProtocolCachePolicy, timeoutInterval:60.0)
		webView.loadRequest(urlRequest)
	}

	func flashMessage(message:String, title:String) {
		guard alertController == nil else {
			// alert already showing.  try again soon.
			let delayTime = DispatchTime.now() + Double(Int64(0.5 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)
			DispatchQueue.main.asyncAfter(deadline: delayTime, execute: {
				self.flashMessage(message: message, title: title)
			})
			return
		}

		guard view.superview != nil else {
			print("Left webview.  Skipping title: \(title), message: \(message)")
			return
		}

		alertController = UIAlertController(title: title, message: message, preferredStyle: .actionSheet)

		present(alertController, animated: true) {
			let delayTime = DispatchTime.now() + Double(Int64(1.0 * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)
			DispatchQueue.main.asyncAfter(deadline: delayTime, execute: {
				self.dismiss(animated: true) {
					self.alertController = nil
				}
			})
		}
	}

	// MARK: UIWebViewDelegate

	func webViewDidStartLoad(_ webView: UIWebView) {
		// Note: we would like to clear network stats before each new page,
		// but this can be called several times per page (once per frame).
		// Outside the scope of this demo, one could accumulate stats across
		// across frames and store those stats for each URL loaded.
	}

	func webViewDidFinishLoad(_ webView: UIWebView) {
		guard let url = url else {
			print("URL not set")
			return
		}

		if !webView.isLoading {
			loadTimeElapsed = Date.timeIntervalSinceReferenceDate - startTime
			flashMessage(message: url.absoluteString, title: String(format: "%.2fs", loadTimeElapsed))
			webDelegate?.loadedWebSiteName(siteName: (webView.request?.mainDocumentURL?.absoluteString)!)
		}
	}

    func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
        flashMessage(message: "Cannot load page", title:"Offline")
    }

}
